beater/gpfs.go

```diff
@@ -89,56 +126,142 @@ func parseCertainInt(s string) int64 {
+	identifier := fields[identifierFieldLocation]
+	if fields[headerFieldLocation] == "HEADER" {
+		fieldMap := parseGpfsHeaderFields(fields)
+		headerMap[identifier] = fieldMap
+		return true
+	}
+	return false
+}
+
+// parseGpfsYOutput parses data produced by a GPFS command using the -Y flag
+func parseGpfsYOutput(
+	prefixFieldlocation int,
+	identifierFieldLocation int,
+	headerFieldLocation int,
+	prefix string,
+	output string,
+	fn parseCallBack) ([]interface{}, error) {
 
 	lines := strings.Split(output, "\n")
-	fieldMap := parseMmRepQuotaHeader(lines[0])
+	var headerMap = make(map[string](map[string]int))
+
+	result := make([]interface{}, 0, len(lines))
 
-	var qs = make([]QuotaInfo, 0, 100000)
+	for _, line := range lines {
 
-	for _, line := range lines[1:] {
+		// ignore empty lines
 		if line == "" {
 			continue
 		}
+
+		// we need to get the fields since we need to know the identifier in order to look up the correct field names
 		fields := strings.Split(line, ":")
-		qi := QuotaInfo{
-			filesystem: fields[fieldMap["filesystemName"]],
-			fileset:    fields[fieldMap["filesetname"]],
-			kind:       fields[fieldMap["quotaType"]],
-			entity:     fields[fieldMap["name"]],
-			blockUsage: parseCertainInt(fields[fieldMap["blockUsage"]]),
-			blockSoft:  parseCertainInt(fields[fieldMap["blockQuota"]]),
-			blockHard:  parseCertainInt(fields[fieldMap["blockLimit"]]),
-			blockDoubt: parseCertainInt(fields[fieldMap["blockInDoubt"]]),
-			blockGrace: fields[fieldMap["blockGrace"]],
-			filesUsage: parseCertainInt(fields[fieldMap["filesUsage"]]),
-			filesSoft:  parseCertainInt(fields[fieldMap["filesQuota"]]),
-			filesHard:  parseCertainInt(fields[fieldMap["filesLimit"]]),
-			filesDoubt: parseCertainInt(fields[fieldMap["filesInDoubt"]]),
-			filesGrace: fields[fieldMap["filesGrace"]],
+
+		// ignore all weird lines :)
+		if fields[prefixFieldlocation] != prefix {
+			continue
 		}
-		if qi.kind == "FILESET" {
-			qi.fileset = qi.entity // filesets have no name, and we need to have a link between FILESET and USR quota
+
+		// there may be multiple HEADER lines so we need to gather them here (which is ugly, granted)
+		// we then also already have the line identifier, so no need to get it from the HEADER parsing
+		if updateHeaderMap(headerFieldLocation, identifierFieldLocation, headerMap, fields) {
+			continue // we updated the map, so we can skip the remainder for this header line
 		}
-		qs = append(qs, qi)
+		identifier := fields[identifierFieldLocation]
+		fieldMap := headerMap[identifier]
+		info := fn(fields, fieldMap)
+		result = append(result, info)
 	}
-	return qs, nil
+
+	return result, nil
 }
 
-// parseMmRepQuotaHeader builds a map of the field names and the corresponding index
-func parseMmRepQuotaHeader(header string) map[string]int {
+// parseMmLsFs returns the different devices in the GPFS cluster
+func parseMmLsFs(output string) ([]string, error) {
+	var prefixFieldlocation = 0
```

Comment:
why are these 3 passed?
and if you have to, why not hardcode 0,1,2 on the function call. using the variables makes the function call really really long.